<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Test motif enrichment using a background set of sequences — testMotifEnrich • motifTestR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Test motif enrichment using a background set of sequences — testMotifEnrich"><meta name="description" content="Test for motif enrichment within a set of sequences using a background set to
derive a NULL hypothesis"><meta property="og:description" content="Test for motif enrichment within a set of sequences using a background set to
derive a NULL hypothesis"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">motifTestR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/motifAnalysis.html">Basic Motif Analysis</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/smped/motifTestR" aria-label="View on github"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Test motif enrichment using a background set of sequences</h1>
      <small class="dont-index">Source: <a href="https://github.com/smped/motifTestR/blob/gh-actions/R/testMotifEnrich.R" class="external-link"><code>R/testMotifEnrich.R</code></a></small>
      <div class="d-none name"><code>testMotifEnrich.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Test for motif enrichment within a set of sequences using a background set to
derive a NULL hypothesis</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">testMotifEnrich</span><span class="op">(</span></span>
<span>  <span class="va">pwm</span>,</span>
<span>  <span class="va">stringset</span>,</span>
<span>  <span class="va">bg</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"iteration"</span>,</span>
<span>  model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"quasipoisson"</span>, <span class="st">"hypergeometric"</span>, <span class="st">"poisson"</span>, <span class="st">"iteration"</span><span class="op">)</span>,</span>
<span>  sort_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  mc.cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-pwm">pwm<a class="anchor" aria-label="anchor" href="#arg-pwm"></a></dt>
<dd><p>A Position Weight Matrix or list of PWMs</p></dd>


<dt id="arg-stringset">stringset<a class="anchor" aria-label="anchor" href="#arg-stringset"></a></dt>
<dd><p>An XStringSet with equal sequence widths</p></dd>


<dt id="arg-bg">bg<a class="anchor" aria-label="anchor" href="#arg-bg"></a></dt>
<dd><p>An XStringSet with the same sequence widths as the test XStringset</p></dd>


<dt id="arg-var">var<a class="anchor" aria-label="anchor" href="#arg-var"></a></dt>
<dd><p>A column in the mcols element of bg, usually denoting an iteration
number</p></dd>


<dt id="arg-model">model<a class="anchor" aria-label="anchor" href="#arg-model"></a></dt>
<dd><p>The model used for analysis</p></dd>


<dt id="arg-sort-by">sort_by<a class="anchor" aria-label="anchor" href="#arg-sort-by"></a></dt>
<dd><p>Column to sort results by</p></dd>


<dt id="arg-mc-cores">mc.cores<a class="anchor" aria-label="anchor" href="#arg-mc-cores"></a></dt>
<dd><p>Passed to <a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Passed to <a href="getPwmMatches.html">getPwmMatches</a> or <a href="countPwmMatches.html">countPwmMatches</a></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.frame with columns: <code>sequences</code>, <code>matches</code>, <code>expected</code>, <code>enrichment</code>,
and <code>p</code>, with additional columns <code>Z</code>, <code>est_bg_rate</code> (Poisson), <code>odds_ratio</code>
(Hypergeometric) or <code>Z</code>, <code>sd_bg</code>, <code>n_iter</code> and <code>iter_p</code> (Iterations).
The numbers of sequences and matches refer to the test set of sequences,
whilst expected is the expected number of matches under the Poisson or
iterative null distribution. The ratio of matches to expected is given as
<code>enrichment</code>, along with the Z score and p-value. Whilst the Z score is only
representative under the Poisson model, it is used to directly estimate
p-values under the iterative approach.
Under this latter approach, the sd of the matches found in the background
sequences is also given, along with the number of iterations and the p-values
from permutations testing the one-sided hypothesis hypothesis for enrichment.</p>
<p>It may also be worth noting that if producing background sequences using
<a href="makeRMRanges-methods.html">makeRMRanges</a> with <code>replace = TRUE</code> and <code>force_ol = TRUE</code>, the
iterative model corresponds to a bootstrap, given that the test sequences
will overlap the background sequences and background ranges are able to be
sampled with replacement.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function offers four alternative models for assessing the enrichment of
a motif within a set of sequences, in comparison to a background set of
sequences.
Selection of the BG sequences plays an important role and, in conjunction
with the question being addressed, determines the most appropriate model to
use for testing, as described below.</p>
<p>It should also be noted that the larger the BG set of sequences, the larger
the computational burden, and results can take far longer to return. For
many millions of background sequences, this may run beyond an hour</p><div class="section">
<h3 id="descriptions-of-models-and-use-cases">Descriptions of Models and Use Cases<a class="anchor" aria-label="anchor" href="#descriptions-of-models-and-use-cases"></a></h3>

<div class="section">
<h4 id="hypergeometric-tests">Hypergeometric Tests<a class="anchor" aria-label="anchor" href="#hypergeometric-tests"></a></h4>


<p>Hypergeometric tests are best suited to the use case where the test set of
sequences represents a subset of a larger set, with a specific feature or
behaviour, whilst the BG set may be the remainder of the set without that
feature. For example, the test set may represent ChIP-Seq binding sites
where signal changes in response to treatment, whilst the BG set represents
the sites where no changed signal was observed. Testing is one-sided, for
enrichment of motifs within the test set.</p>
<p>Due to these relatively smaller sized datasets, setting
model = "hypergeometric", will generally return results quickly</p>
</div>

<div class="section">
<h4 id="poisson-tests">Poisson Tests<a class="anchor" aria-label="anchor" href="#poisson-tests"></a></h4>


<p>This approach requires a set of background sequences which should be much
larger than the test set of sequences.
The parameters for a Poisson model are estimated in a per-sequence manner on
the set of BG sequences, and the observed rate of motif-matches within the
test set is then tested using <a href="https://rdrr.io/r/stats/poisson.test.html" class="external-link">poisson.test</a>.</p>
<p>This approach assumes that all matches follow a Poisson distribution, which
is often true, but data can also be overdispersed. Given that this model can
also return results relatively quickly, is it primarily suitable for data
exploration, such as quickly checking for expected behaviours, but not for
final results.</p>
</div>

<div class="section">
<h4 id="quasi-poisson-test">Quasi-Poisson Test<a class="anchor" aria-label="anchor" href="#quasi-poisson-test"></a></h4>


<p>The quasipoisson model allows for over-dispersion and will return more
conservative results than using the standard Poisson model.
Under the method currently implemented here, BG sequences should be divided
into blocks (i.e. iterations), identical in size to the test set of sequences.
Model parameters are estimated per iteration across the BG set of sequences,
with the rate of matches in the test set being compared against these blocks.
This ensures more conservative results that if analysing test and bg
sequences as collections of individual sequences.</p>
<p>It is expected that the BG set will matched for the features of interest and
chosen using <a href="makeRMRanges-methods.html">makeRMRanges</a> with a large number of iterations, e.g.
n_iter = 1000.
Due to this parameterisation, quasipoisson approaches can be computationally
time-consuming, as this is effectively an iterative approach.</p>
</div>

<div class="section">
<h4 id="iteration">Iteration<a class="anchor" aria-label="anchor" href="#iteration"></a></h4>


<p>Setting the model as "iteration" performs a non-parametric analysis, with
the exception of returning Z-scores under the Central Limit Theorem.
Mean and SD of matches is found for each iteration, and used to return Z
scores, with p-values returned from both a Z-test and from comparing
observed values directly to sampled values obtained from the BG sequences.
Sampled values are calculated directly and as such, are limited in precision.</p>
<p>As for the QuasiPoisson model, a very large number of iterations is expected
to be used, to ensure the CLT holds, again making this a computationally
demanding test.
Each iteration/block is expected to be identically-sized to the test set,
and matched for any features as appropriate using <code><a href="makeRMRanges-methods.html">makeRMRanges()</a></code>.</p>
</div>


</div>

    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="makeRMRanges-methods.html">makeRMRanges()</a></code>, <code><a href="getPwmMatches.html">getPwmMatches()</a></code>, <code><a href="countPwmMatches.html">countPwmMatches()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Load the example peaks &amp; the sequences</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ar_er_peaks"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ar_er_seq"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sq</span> <span class="op">&lt;-</span> <span class="fu">seqinfo</span><span class="op">(</span><span class="va">ar_er_peaks</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## Now sample size-matched ranges 10 times larger. In real-world analyses,</span></span></span>
<span class="r-in"><span><span class="co">## this set should be sampled as at least 1000x larger, ensuring features</span></span></span>
<span class="r-in"><span><span class="co">## are matched to your requirements. This example masks regions with known N</span></span></span>
<span class="r-in"><span><span class="co">## content, including centromeres &amp; telomeres</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"hg19_mask"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">305</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bg_ranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="makeRMRanges-methods.html">makeRMRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">ar_er_peaks</span>, <span class="fu">GRanges</span><span class="op">(</span><span class="va">sq</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, exclude <span class="op">=</span> <span class="va">hg19_mask</span>, n_iter <span class="op">=</span> <span class="fl">10</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Convert ranges to DNAStringSets</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="va">BSgenome.Hsapiens.UCSC.hg19</span></span></span>
<span class="r-in"><span><span class="va">bg_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html" class="external-link">getSeq</a></span><span class="op">(</span><span class="va">genome</span>, <span class="va">bg_ranges</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Test for enrichment of the ESR1 motif</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ex_pfm"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">esr1</span> <span class="op">&lt;-</span> <span class="va">ex_pfm</span><span class="op">$</span><span class="va">ESR1</span></span></span>
<span class="r-in"><span><span class="fu">testMotifEnrich</span><span class="op">(</span><span class="va">esr1</span>, <span class="va">ar_er_seq</span>, <span class="va">bg_seq</span>, model <span class="op">=</span> <span class="st">"poisson"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   sequences matches expected enrichment        Z            p          fdr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       849      22      0.5         44 30.40559 1.315112e-28 1.315112e-28</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    est_bg_rate</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 0.0005889282</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Test all motifs</span></span></span>
<span class="r-in"><span><span class="fu">testMotifEnrich</span><span class="op">(</span><span class="va">ex_pfm</span>, <span class="va">ar_er_seq</span>, <span class="va">bg_seq</span>, model <span class="op">=</span> <span class="st">"poisson"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       sequences matches expected enrichment         Z            p          fdr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ZN143       849      21      0.0        Inf       Inf 0.000000e+00 0.000000e+00</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FOXA1       849     113     25.5   4.431373 17.327582 4.223239e-37 1.055810e-36</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ESR1        849      22      0.5  44.000000 30.405592 1.315112e-28 2.191853e-28</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ANDR        849       8      1.4   5.714286  5.578018 1.065480e-04 1.331850e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ZN281       849      14      6.4   2.187500  3.004164 7.912704e-03 7.912704e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        est_bg_rate</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ZN143 0.0000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FOXA1 0.0300353357</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ESR1  0.0005889282</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ANDR  0.0016489988</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ZN281 0.0075382803</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/smped" class="external-link">Stevie Pederson</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

